expr/js_eval.go

```diff
@@ -0,0 +1,211 @@
+	evaluatedValues := []otto.Value{}
+	if result.IsObject() {
+		switch result.Object().Class() {
+		case "GoSlice", "Array":
+			for _, key := range result.Object().Keys() {
+				// This should always work, as we just asked for the available keys.
+				element, err := result.Object().Get(key)
+				if err != nil {
+					return resultValues, err
+				}
+
+				evaluatedValues = append(evaluatedValues, element)
+			}
+		}
+	} else {
+		return resultValues, errors.New("expected array input, but source does not contain multiple values")
+	}
+
+	// We parsed our JS successfully, and have multiple values, as expected.
+	// Now parse each nested value and return the final slice.
+	for _, evaluatedValue := range evaluatedValues {
+		resultValue, err := EvaluateResultType[ReturnType](ctx, evaluatedValue)
+		if err != nil {
+			return resultValues, nil
+		}
+		resultValues = append(resultValues, resultValue)
+	}
+
+	return resultValues, nil
+}
+
+func EvaluateSingleValue[ReturnType any](ctx context.Context, source string, subject any) (ReturnType, error) {
+	result, err := EvaluateJavascript(ctx, source, subject)
+	if err != nil {
+		// This is an expected situation: lots of Javascript code will fail depending on the
+		// input it receives.
+
+		// For now, while we do nothing with the errors (not what we'll always do) we should
+		// log only with an event that we can easily search for.
+		fmt.Println("Error when evaluating javascript", err)
+
+		// When failing, treat the result as if it were an undefined value.
+		result = otto.UndefinedValue()
+	}
+
+	resultValue, err := EvaluateResultType[ReturnType](ctx, result)
+	if err != nil {
+		return resultValue, err
+	}
+
+	return resultValue, nil
+}
+
+func EvaluateResultType[ReturnType any](ctx context.Context, result otto.Value) (ReturnType, error) {
+	var resultValue ReturnType
+	var ok bool
+	switch {
+	case result.IsBoolean():
+		resultBool, err := result.ToBoolean()
+		if err != nil {
+			return resultValue, err
+		}
+
+		// This is a pattern we'll employ in each of the checks below
+		// to see if our result value matches the expected ReturnType.
+		// It's slightly gross, but does the trick.
+		typeAgnosticResult := any(resultBool)
+
+		// If OK, this is supported by Bool.
+		resultValue, ok := typeAgnosticResult.(ReturnType)
+
+		// In bool's case only, if not ok, try the value again as a string.
+		if !ok {
+			boolValue := fmt.Sprintf("%v", resultBool)
+			typeAgnosticResult := any(boolValue)
+			resultValue, ok = typeAgnosticResult.(ReturnType)
+			if !ok {
+				return resultValue, fmt.Errorf("could not convert result of bool to %T", resultValue)
+			}
+		}
+
+		return resultValue, nil
+
+	case result.IsNumber():
+		resultInt, err := strconv.Atoi(fmt.Sprintf("%v", result))
+		if err != nil {
+			return resultValue, err
+		}
+
+		typeAgnosticResult := any(resultInt)
+
+		// If OK, this is supported by Number.
+		resultValue, ok = typeAgnosticResult.(ReturnType)
+		if !ok {
```

Comment:
I wonder if we should do the bool trick here too, so if you return a number we still can parse as a string?

Reply:
agree